File: examples/example-54/js/features/step_definitions/messaging.js
const { When, Then } = require('@cucumber/cucumber');
const { getLastMessage } = require('../../src/messaging');
const assert = require('assert');
const promiseRetry = require('promise-retry');

When('a message {string} is sent to the queue {string}', async function (msg, q) {
    await this.amqp.channel.sendToQueue(q, Buffer.from(msg));
});

Then('the last message received is {string}', function (expectedMessage) {
    return promiseRetry(async function (retry) {
        try {
            assert.strictEqual(getLastMessage(), expectedMessage);
        } catch (ex) { retry(); }
    }, {
        retries: 10,
        factor: 2,
        minTimeout: 100,
        maxTimeout: 2000,
    });
});

When('a message {string} is sent to the topic {string}', async function (msg, t) {
    await this.amqp.channel.publish('topic_exchange', t, Buffer.from(msg));
});

Then('a message {string} should be received by the topic {string}', function (m, t) {
    return promiseRetry((retry) => {
        try {
            assert.ok(this.amqp?.topics?.[t]);
            assert.ok(this.amqp.topics[t].includes(m));
        } catch (ex) { retry(); }
    }, {
        retries: 10,
        factor: 2,
        minTimeout: 100,
        maxTimeout: 2000,
    });
});

File: examples/example-54/ruby/features/step_definitions/messaging.rb
require('rspec')
require('retries')

When('a message {string} is sent to the queue {string}') do |message, queue_name|
    queue = @channel.queue(queue_name)
    queue.publish(message)
end

Then('the last message received is {string}') do |message|
    with_retries(
        max_tries: 10,
        rescue: [RSpec::Expectations::ExpectationNotMetError],
        base_sleep_seconds: 0.1,
        max_sleep_seconds: 2.0
    ) do
        expect($last_message).to eq(message)
    end
end

When('a message {string} is sent to the topic {string}') do |message, topic|
    @topic_exchange.publish(message, routing_key: topic)
end

Then('a message {string} should be received by the topic {string}') do |msg, t|
    with_retries(
        max_tries: 10,
        rescue: [RSpec::Expectations::ExpectationNotMetError],
        base_sleep_seconds: 0.1,
        max_sleep_seconds: 2.0
    ) do
        msgs = @topic_messages
               .select { |m| m[:topic] == t }
               .map { |m| m[:body] }
        expect(msgs).to include(msg)
    end
end
